engine/evd.ml

```diff
@@ -828,35 +833,20 @@ let drop_all_defined d = { d with defn_evars = EvMap.empty }
 (* spiwack: not clear what folding over an evar_map, for now we shall
     simply fold over the inner evar_map. *)
 let fold f d a =
+  let f evk evi accu = f evk (EvarInfo evi) accu in
   EvMap.fold f d.defn_evars (EvMap.fold f d.undf_evars a)
 
 let fold_undefined f d a = EvMap.fold f d.undf_evars a
 
+type map = { map : 'r. Evar.t -> 'r evar_info -> 'r evar_info }
+
 let raw_map f d =
-  let f evk info =
-    let ans = f evk info in
-    let () = match info.evar_body, ans.evar_body with
-    | Evar_defined _, Evar_empty
-    | Evar_empty, Evar_defined _ ->
-      anomaly (str "Unrespectful mapping function.")
-    | _ -> ()
-    in
-    ans
-  in
+  let f (type a) evk (info : a evar_info) = f.map evk info in
   let defn_evars = EvMap.Smart.mapi f d.defn_evars in
   let undf_evars = EvMap.Smart.mapi f d.undf_evars in
```

Comment:
```suggestion
  let defn_evars = EvMap.Smart.mapi f.map d.defn_evars in
  let undf_evars = EvMap.Smart.mapi f.map d.undf_evars in
```

Reply:
This doesn't typecheck on my machine for some reason.

Reply:
It works on my machine, let's ask CI

Reply:
Hmm, weird, now it works for me locally. Maybe I did something wrong last time I tried.